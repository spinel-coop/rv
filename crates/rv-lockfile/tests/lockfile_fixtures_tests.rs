use rv_lockfile::{parse_lockfile, parse_lockfile_strict, ParseError};
use insta::assert_debug_snapshot;
use std::fs;
use std::path::PathBuf;

/// Helper function to load a test fixture
fn load_fixture(name: &str) -> String {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/fixtures");
    path.push(format!("{}.gemfile.lock", name));
    
    fs::read_to_string(&path)
        .unwrap_or_else(|_| panic!("Failed to read fixture: {}", path.display()))
}

/// Test a lockfile fixture and create a snapshot
fn test_fixture(name: &str) {
    let content = load_fixture(name);
    let parser = parse_lockfile(&content).unwrap();
    assert_debug_snapshot!(format!("fixture_{}", name), parser);
}

/// Test a lockfile fixture with strict mode and create a snapshot
fn test_fixture_strict(name: &str) {
    let content = load_fixture(name);
    let parser = parse_lockfile_strict(&content).unwrap();
    assert_debug_snapshot!(format!("fixture_{}_strict", name), parser);
}

/// Test a lockfile fixture that should fail parsing
fn test_fixture_error(name: &str) {
    let content = load_fixture(name);
    let result = parse_lockfile(&content);
    assert!(result.is_err());
    assert_debug_snapshot!(format!("fixture_{}_error", name), result.unwrap_err());
}

#[test]
fn test_empty_lockfile() {
    test_fixture("empty");
}

#[test]
fn test_minimal_lockfile() {
    test_fixture("minimal");
}

#[test]
fn test_git_source_lockfile() {
    test_fixture("git_source");
}

#[test]
fn test_gem_source_lockfile() {
    test_fixture("gem_source");
}

#[test]
fn test_multi_platform_lockfile() {
    test_fixture("multi_platform");
}

#[test]
fn test_path_source_lockfile() {
    test_fixture("path_source");
}

#[test]
fn test_checksum_enabled_lockfile() {
    test_fixture("checksum_enabled");
}

#[test]
fn test_rails_app_lockfile() {
    test_fixture("rails_app");
}

// Test the same fixtures in strict mode
#[test]
fn test_empty_lockfile_strict() {
    test_fixture_strict("empty");
}

#[test]
fn test_minimal_lockfile_strict() {
    test_fixture_strict("minimal");
}

#[test]
fn test_git_source_lockfile_strict() {
    test_fixture_strict("git_source");
}

// Create inline test for merge conflict since it can't be a file
#[test]
fn test_merge_conflict_detection() {
    let conflict_content = r#"
GEM
  remote: https://rubygems.org/
  specs:
<<<<<<< HEAD
    gem-a (1.0.0)
=======
    gem-b (2.0.0)
>>>>>>> feature-branch

PLATFORMS
  ruby
"#;
    
    let result = parse_lockfile(conflict_content);
    assert!(result.is_err());
    
    if let Err(ParseError::MergeConflict { line }) = result {
        assert_eq!(line, 5); // Line with <<<<<<< HEAD
    } else {
        panic!("Expected MergeConflict error, got: {:?}", result);
    }
}