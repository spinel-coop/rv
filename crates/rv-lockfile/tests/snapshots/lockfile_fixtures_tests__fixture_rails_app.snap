---
source: crates/rv-lockfile/tests/lockfile_fixtures_tests.rs
expression: parser
---
LockfileParser {
    sources: [
        Gem(
            GemSource {
                remotes: [
                    "https://rubygems.org/",
                ],
                specs: [
                    "actioncable",
                    "actionmailbox",
                    "actionmailer",
                    "rails",
                    "bootsnap",
                    "puma",
                ],
            },
        ),
    ],
    specs: {
        "actioncable (7.0.4)": LazySpecification {
            name: "actioncable",
            version: Version {
                major: 7,
                minor: 0,
                patch: 4,
            },
            platform: Ruby,
            source: None,
            dependencies: [],
            checksum: None,
            required_ruby_version: None,
            required_rubygems_version: None,
        },
        "actionmailbox (7.0.4)": LazySpecification {
            name: "actionmailbox",
            version: Version {
                major: 7,
                minor: 0,
                patch: 4,
            },
            platform: Ruby,
            source: None,
            dependencies: [],
            checksum: None,
            required_ruby_version: None,
            required_rubygems_version: None,
        },
        "actionmailer (7.0.4)": LazySpecification {
            name: "actionmailer",
            version: Version {
                major: 7,
                minor: 0,
                patch: 4,
            },
            platform: Ruby,
            source: None,
            dependencies: [],
            checksum: None,
            required_ruby_version: None,
            required_rubygems_version: None,
        },
        "rails (7.0.4)": LazySpecification {
            name: "rails",
            version: Version {
                major: 7,
                minor: 0,
                patch: 4,
            },
            platform: Ruby,
            source: None,
            dependencies: [],
            checksum: None,
            required_ruby_version: None,
            required_rubygems_version: None,
        },
        "bootsnap (1.13.0)": LazySpecification {
            name: "bootsnap",
            version: Version {
                major: 1,
                minor: 13,
                patch: 0,
            },
            platform: Ruby,
            source: None,
            dependencies: [],
            checksum: None,
            required_ruby_version: None,
            required_rubygems_version: None,
        },
        "puma (5.6.5)": LazySpecification {
            name: "puma",
            version: Version {
                major: 5,
                minor: 6,
                patch: 5,
            },
            platform: Ruby,
            source: None,
            dependencies: [],
            checksum: None,
            required_ruby_version: None,
            required_rubygems_version: None,
        },
    },
    dependencies: {
        "bootsnap": Dependency {
            name: "bootsnap",
            requirements: [],
            platforms: [],
            source: None,
            pinned: false,
        },
        "puma": Dependency {
            name: "puma",
            requirements: [],
            platforms: [],
            source: None,
            pinned: false,
        },
        "rails": Dependency {
            name: "rails",
            requirements: [],
            platforms: [],
            source: None,
            pinned: false,
        },
    },
    platforms: [
        Ruby,
        Specific {
            cpu: "x86_64",
            os: "linux",
            version: None,
        },
    ],
    bundler_version: Some(
        Version {
            major: 2,
            minor: 3,
            patch: 22,
        },
    ),
    ruby_version: Some(
        "3.1.2p20",
    ),
    checksums_enabled: false,
    checksums: {},
    strict: false,
}
